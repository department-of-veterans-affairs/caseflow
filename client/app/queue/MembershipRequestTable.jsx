import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { DoubleArrowIcon } from '../components/icons/DoubleArrowIcon';
import moment from 'moment';
import DropdownButton from '../components/DropdownButton';
import Table from '../components/Table';
import { constant, cloneDeep } from 'lodash';
import Pagination from '../components/Pagination/Pagination';

const MembershipRequestTable = (props) => {

  const { requests = [] } = props;

  const REQUESTS_PER_PAGE = 10;

  const initializePaginatedData = (membershipRequests) => {
    const paginatedData = [];

    for (let i = 0; i < membershipRequests.length; i += REQUESTS_PER_PAGE) {
      paginatedData.push(membershipRequests.slice(i, i + REQUESTS_PER_PAGE));
    }

    return paginatedData;
  };

  // State variable to store the expanded state. It is an object of the form { [request.id]: boolean }
  const [expanded, setExpanded] = useState({});
  const [currentPage, setCurrentPage] = useState(0);

  // Set the row objects for the table based on if the request has a note or not.
  // If it has a note and is expanded, add an additional row in the table.
  // The expanded is set based on the expanded react state hook and the button click
  const getRowObjects = (rows) => {

    const updatedRequests = rows.reduce((acc, request) => {
      acc.push(request);
      if (request.note && expanded[request.id]) {
        acc.push({ ...request, hasNote: true });
      }

      return acc;
    }, []);

    return updatedRequests;
  };

  // Since we are modifying the current page paginated rows instead of all of them now
  const getPaginatedRowObjects = (rows) => {
    // Guard clause for empty rows.
    if (!rows) {
      return [];
    }
    const updatedRequests = rows.reduce((acc, request) => {
      // If it does not have a note add it to the list. If it has a note skip it to avoid duplicates.
      if (!request.hasNote) {
        acc.push(request);
      }

      if ((request.note && expanded[request.id]) && !request.hasNote) {
        acc.push({ ...request, hasNote: true });
      }

      return acc;
    }, []);

    return updatedRequests;
  };

  // Create a state variable to hold the paginated requests.
  const [paginatedRequests, setPaginatedRequests] = useState(initializePaginatedData(getRowObjects(requests)));

  const updatePaginatedData = () => {
    const tempArray = cloneDeep(paginatedRequests);

    tempArray[currentPage] = getPaginatedRowObjects(tempArray[currentPage]);
    setPaginatedRequests(tempArray);
  };

  // Updates the current page pagination data
  useEffect(() => {
    updatePaginatedData();
  }, [expanded]);

  // Update the state if the requests prop changes.
  useEffect(() => {
    setPaginatedRequests(initializePaginatedData(getRowObjects(requests)));
  }, [requests]);

  const toggleExpanded = (id) => {
    setExpanded({
      ...expanded,
      [id]: !expanded[id],
    });
  };

  const dropdownOptions = [
    {
      title: 'Approve',
      action: 'approved',
    },
    {
      title: 'Deny',
      action: 'denied',
    },
  ];

  const buildDropdownActions = (request) => {
    return dropdownOptions.map((obj) => {
      return { ...obj, value: `${request.id}-${obj.action}` };
    });
  };

  const MembershipRequestDropDown = () => <div>
    <span> Select action </span>
    <DoubleArrowIcon />
  </div>;

  // The normal column definitions for the request table row that is not the additional request note row.
  const normalColumnsDefinitions = [
    {
      header: 'User name',
      valueFunction: (request) => request.userNameWithCssId
    },
    {
      header: 'Date requested',
      valueFunction: (request) => moment(request.requestedDate).format('MM/DD/YYYY')
    },
    {
      header: 'Actions',
      valueFunction: (request) => {
        return <DropdownButton
          lists={buildDropdownActions(request)}
          label="Request actions"
          onClick={props.membershipRequestActionHandler}
        >
          <MembershipRequestDropDown />
        </DropdownButton>;
      }
    },
    {
      header: '',
      valueFunction: (request) => {
        if (request.note) {
          return <button style={{ backgroundColor: 'inherit' }}
            onClick={() => toggleExpanded(request.id)}
            className="usa-accordion-button"
            aria-expanded={Boolean(expanded[request.id])}
            aria-label="Show note">
          </button>;
        }

        return '';

      },
    },
  ];

  // If it is a note row, return the request's note text and set the colspan to 4.
  const noteColumnDefinition = [
    {
      header: '',
      valueFunction: (request) => {
        return <div className="expanded-note">
          <strong>REQUEST NOTE:</strong>
          <p>{request.note}</p>
        </div>;
      },
      span: constant(4),
    },
  ];

  // The column defintion will change for the table cell for each row depending on whether the row/request
  // has the attribute hasNote which is set in the row objects based on the expanded state.
  const columnDefinitions = (row) => {

    if (row && row.hasNote) {
      return noteColumnDefinition;
    }

    return normalColumnsDefinitions;
  };

  // Conditionally set the row class name for css for note rows.
  const setRowClassNames = (rowObject) => {
    return rowObject.hasNote ? 'membership-request-expanded-note-row' : 'membership-request-row';
  };

  // Only count the rows without the hasNote property key since those aren't membership requests, but notes.
  let currentRequestsCount = 0;

  if (paginatedRequests[currentPage]) {
    currentRequestsCount = paginatedRequests[currentPage].filter((obj) => !obj.hasNote).length;
  }

  // Create a function for the pagination so it can be easily added above and below the table.
  const MembershipRequestPagination = () => {
    return <Pagination
      pageSize={REQUESTS_PER_PAGE}
      currentPage={currentPage + 1}
      currentCases={currentRequestsCount}
      totalPages={paginatedRequests.length}
      totalCases={requests.length}
      updatePage={(newPage) => setCurrentPage(newPage)}
    />;
  };

  return <>
    <h2>{`View ${requests.length} pending requests`}</h2>
    <MembershipRequestPagination />
    <Table
      className="membership-request-table"
      columns={columnDefinitions}
      rowObjects={paginatedRequests[currentPage] || []}
      rowClassNames={setRowClassNames}
      getKeyForRow={(_, { hasNote, id }) => hasNote ? `${id}-note` : `${id}`}
    />
    <MembershipRequestPagination />
  </>;
};

MembershipRequestTable.propTypes = {
  requests: PropTypes.array,
  enablePagination: PropTypes.bool,
  membershipRequestActionHandler: PropTypes.func.isRequired,
};

export default MembershipRequestTable;
