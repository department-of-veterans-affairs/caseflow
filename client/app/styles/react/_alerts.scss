
.usa-alert {

  &.fixed {
    position: fixed;
    top: 0;
    margin-top: 0;
    z-index: 999;
    left: 50%;
    transform: translate(-50%, 0);
  }

  // Remove after USWDS V.2.0
  &.usa-alert-slim {
    background-size: 4rem;

    .usa-alert-body {
      padding-left: 3.5rem;
    }

    .usa-alert-text {
      font-size: 2rem;
    }
  }

  .usa-alert-text {
    line-height: 1.5;
  }
}
