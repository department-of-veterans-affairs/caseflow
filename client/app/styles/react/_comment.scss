.comment-control-button-container {
  display: flex;
  width: 100%;
  padding: 3px 0;
  justify-content: space-between;
}

.comment-save-button-container {
  display: inline-block;
  width: 100%;
  margin-top: 5px;
  margin-bottom: 17px;
}

.comment-control-button {
  padding: 0;

  &.horizontal {
    text-align: left;
  }
}

.comment-control-button-divider {
  padding: 0 10px;
}

.comment-container-selected {
  box-shadow: 3px 3px 5px $color-gray-light;
}

.comment-container {
  word-wrap: break-word;
  word-break: break-word;
  background-color: $color-white;
  padding: 15px;
  cursor: pointer;

  &:not(.comment-horizontal-container) {
    margin-bottom: 15px;
    border-radius: 3px;
    border: 1px solid $color-gray-lighter;
  }

  &.comment-horizontal-container {
    padding: 25px 25px 25px 40px;
  }
}

// Max width is enforced so the size of this comment box
// fits with existing textareas in style guide
.comment-size-container {
  max-width: 430px;

  .cf-form-textarea {
    margin-bottom: 20px;
  }
}

.comment-horizontal-container {
  display: inline-block;
  width: 100%;
}

.comment-horizontal-spacing {
  margin-right: 49px;
}

.commentIcon-container {
  position: absolute;
  z-index: 10;
  // These dimensions need to match the comment icon SVG dimensions.
  width: 40px;
  height: 40px;
}

