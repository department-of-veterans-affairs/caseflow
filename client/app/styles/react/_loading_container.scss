.loadingContainer-positioning {
  position: absolute;
  width: 100%;
  height: 100%;
}

.loadingContainer-table {
  width: 100%;
  height: 100%;
  display: table;
}

.loadingContainer-table-cell {
  text-align: center;
  display: table-cell;
  vertical-align: middle;
}

.loadingContainer-content {
  position: relative;
  z-index: 1;

  iframe {
    width: 100%;
    height: 80vh;

    &.form9Viewer {
      height: 500px;
    }
  }
}

.loadingContainer-container {
  position: relative;
  overflow: hidden;
}

.cf-loading-button-text {
  display: inline-block;
  margin-right: 10px;
  line-height: 0;
  position: relative;
}

.cf-loading-message {
  .loading-message-text {
    display: inline-block;
    font-style: italic;
    font-weight: bolder;
  }

}

.cf-loading-button-symbol {
  display: inline-block;
}
