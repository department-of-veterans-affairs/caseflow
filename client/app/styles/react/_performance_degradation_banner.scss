.usa-banner {
  background-color: $color-gold-lightest;
  text-align: center;
}

.usa-banner-inner {
  padding: .8em;
}

.banner-icon {
  display: inline-block;
  vertical-align: text-top;
  // height: 25px;
}

.banner-text {
  padding-left: 6px;
}
