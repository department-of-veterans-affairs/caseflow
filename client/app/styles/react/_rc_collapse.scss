@mixin common() {

  width: 0;
  height: 0;
  font-size: 0;
  line-height: 0;
}

@mixin right($w, $h, $color) {

  border-top: $w solid transparent;
  border-bottom: $w solid transparent;
  border-left: $h solid $color;
}

@mixin bottom($w, $h, $color) {

  border-left: $w solid transparent;
  border-right: $w solid transparent;
  border-top: $h solid $color;
}

// scss-lint:disable IdSelector
#arrow {
  @include common();
  @include right(3px, 4px, $color-gray);
  @include bottom(3px, 4px, $color-gray);
}

.rc-collapse {

  &-anim-active {
    transition: height .2s ease-out;
  }

  & > &-item {
    &:first-child {
      border-top: none;
    }
  }

  & > &-item-disabled > .rc-collapse-header {
    cursor: not-allowed;
    color: $color-gray-light;
    background-color: $color-gray-lighter;
  }

  &-content {
    overflow: hidden;

    &-inactive {
      display: none;
    }
  }

  &-item:last-child {
    > .rc-collapse-content {
      border-radius: 0 0 3px 3px;
    }
  }

  & > &-item-active {
    > .rc-collapse-header {

      .arrow {

        #arrow > {
          @include bottom(3px, 4px, $color-gray);
        }
      }
    }
  }
}

.usa-accordion-content {
  &:not([aria-hidden]) {
    position: relative;
    left: 0;
  }
}
