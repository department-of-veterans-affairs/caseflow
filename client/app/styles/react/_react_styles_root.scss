@import '_comment.scss';
@import '_search_bar.scss';
@import '_performance_degradation_banner.scss';
@import '_rc_collapse.scss';
@import '_loading_container.scss';
@import '_select-dropdown.scss';
@import '_toggle_button.scss';
