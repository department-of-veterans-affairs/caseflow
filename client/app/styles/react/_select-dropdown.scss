$cf-select-bg-color: #fff;
$cf-select-disabled-bg-color: $color-gray-lightest;
$cf-select-disabled-color: #323a45;
$cf-select-border-color: #5b616b;
$cf-select-text-color: #323a45;

.cf-select {
  .cf-select__control {
    background-color: $cf-select-bg-color;
    border: 1px solid $cf-select-border-color;
    border-collapse: separate;
    border-radius: 0;
    border-spacing: 0;
    color: $cf-select-text-color;
    cursor: default;

    &--is-disabled {
      background-color: $cf-select-disabled-bg-color;

      .cf-select__single-value {
        color: $cf-select-disabled-color;
      }

      .cf-select__indicators {
        opacity: 0.35;
      }

      input {
        margin: 0;
      }
    }

    .cf-select__value-container {
      padding-bottom: 0;
      padding-top: 0;

      .cf-select__placeholder {
        color: $cf-select-text-color;
      }

      // Ensure that the "selected" value is aligned correctly
      .cf-select__placeholder,
      .cf-select__single-value {
        position: unset;
        top: unset;
        transform: unset;
      }
    }

    .cf-select__indicators {
      .cf-select__indicator-separator {
        background-color: $cf-select-border-color;
        margin-bottom: 0;
        margin-top: 0;
      }

      .cf-select__indicator {
        svg {
          fill: $cf-select-text-color;
        }
      }
    }
  }

  .cf-select__menu {
    border-radius: 0;

    .cf-select__menu-list {
      .cf-select__option {
        background-color: $color-white;
        color: $color-gray;

        &.cf-select__option--is-focused {
          background-color: $color-gray;
          color: $color-white;
        }
      }
    }
  }

  .usa-input-error & {
    .cf-select--is-disabled {
      input { width: 1px; }
    }
  }
}
