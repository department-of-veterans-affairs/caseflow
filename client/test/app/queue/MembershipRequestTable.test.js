import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { createMockedMembershipRequests, mockedMembershipRequests } from '../../data/membershipRequests';
import MembershipRequestTable from '../../../app/queue/MembershipRequestTable';
import { axe } from 'jest-axe';

describe('MembershipRequestTable', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const setup = (requests = [], actionHandler = jest.fn) => {
    return render(<MembershipRequestTable requests={requests} membershipRequestActionHandler={actionHandler} />);
  };

  it('renders the default state with no requests correctly', () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it('renders the default state with mocked membership requests correctly', () => {
    const { container } = setup(mockedMembershipRequests);

    expect(container).toMatchSnapshot();
  });

  it('passes a11y', async () => {
    const { container } = setup(mockedMembershipRequests);

    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });

  it('shows the request note when the expansion button is selected', async () => {
    const { container } = setup(mockedMembershipRequests);

    // Since there are multiple expansion buttons grab the first button and later the second button
    let button = screen.getAllByRole('button', { name: 'Show note' })[0];

    // The first note text should not be visible on the screen
    await expect(screen.queryByText(mockedMembershipRequests[0].note)).toBeNull();
    // await button.click();
    await fireEvent.click(button);

    button = screen.getAllByRole('button', { name: 'Show note' })[0];

    // These strings are based on the mocked data. This test assumes the first two mocked requests have notes.
    expect(button.getAttribute('aria-expanded')).toBe('true');
    await waitFor(() => expect(screen.getByText(mockedMembershipRequests[0].note)).toBeInTheDocument());
    await expect(screen.queryByText(mockedMembershipRequests[1].note)).toBeNull();

    // Click the button again and check for the text
    // await button.click();
    await fireEvent.click(button);
    button = screen.getAllByRole('button', { name: 'Show note' })[0];
    expect(button.getAttribute('aria-expanded')).toBe('false');
    // await waitFor(() => expect(screen.getByText(mockedMembershipRequests[0].note)).not.toBeInTheDocument());
    await expect(screen.queryByText(mockedMembershipRequests[0].note)).toBeNull();

    // Get the second note expansion button
    button = screen.getAllByRole('button', { name: 'Show note' })[1];

    // await button.click();
    await fireEvent.click(button);

    // Check for the second note text
    await waitFor(() => expect(screen.getByText(mockedMembershipRequests[1].note)).toBeInTheDocument());

    expect(container).toMatchSnapshot();

  });

  it('displays the pagination controls when there are more than 10 membership requests correctly', async () => {
    const extraMembershipRequests = createMockedMembershipRequests(12);
    const userName = 'Testy McGee';
    const lastRequest = {
      id: 999,
      userNameWithCssId: userName,
      requestedDate: '2023-1-5'
    };

    const requests = [mockedMembershipRequests, extraMembershipRequests, lastRequest].flat();

    setup(requests);

    const paginationButton = screen.getAllByRole('button', {
      name: 'Page 2',
    })[0];

    expect(paginationButton).toBeVisible(true);

    expect(screen.getByText(`View ${requests.length} pending requests`)).toBeVisible(true);

    await fireEvent.click(paginationButton);
    await waitFor(() => expect(screen.getByText(userName)).toBeInTheDocument());

  });

  it('should call the membershipRequestActionHandler property with args when Approve is clicked', async () => {
    const actionHandler = jest.fn();

    setup(mockedMembershipRequests, actionHandler);

    const selectActionButton = screen.getAllByRole('button', { name: 'Request actions' })[0];

    selectActionButton.click();
    await screen.getByText('Approve').click();

    expect(actionHandler).toHaveBeenCalledWith('1-approved');
  });

  it('should call the membershipRequestActionHandler property with args when Deny is clicked', async () => {
    const actionHandler = jest.fn();

    setup(mockedMembershipRequests, actionHandler);

    const selectActionButton = screen.getAllByRole('button', { name: 'Request actions' })[0];

    selectActionButton.click();
    await screen.getByText('Deny').click();

    expect(actionHandler).toHaveBeenCalledWith('1-denied');
  });

});

