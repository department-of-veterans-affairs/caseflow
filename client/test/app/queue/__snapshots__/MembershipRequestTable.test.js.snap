// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MembershipRequestTable renders the default state with mocked membership requests correctly 1`] = `
<div>
  <h2>
    View 5 pending requests
  </h2>
  <table
    class="usa-table-borderless membership-request-table"
  >
    <thead
      class=""
    >
      <tr>
        <th
          class=""
          scope="col"
        >
          <span>
            User name
          </span>
        </th>
        <th
          class=""
          scope="col"
        >
          <span>
            Date requested
          </span>
        </th>
        <th
          class=""
          scope="col"
        >
          <span>
            Actions
          </span>
        </th>
        <th
          class=""
          scope="col"
        >
          <span />
        </th>
      </tr>
    </thead>
    <tbody
      class=""
    >
      <tr
        class="membership-request-row"
        id="table-row-0"
      >
        <td
          class=""
          colspan="1"
        >
          John Doe
        </td>
        <td
          class=""
          colspan="1"
        >
          12/05/2022
        </td>
        <td
          class=""
          colspan="1"
        >
          <div
            class="cf-dropdown"
            data-css-fp8ian=""
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              aria-label="Select action"
              aria-pressed="false"
              class="cf-dropdown-trigger usa-button usa-button-secondary"
              data-css-x1q03x=""
              id="dropdown-button"
              role="dropdown-button"
            >
              Select action
            </button>
          </div>
        </td>
        <td
          class=""
          colspan="1"
        >
          <button
            aria-expanded="false"
            aria-label="Show note"
            class="usa-accordion-button"
            style="background-color: inherit;"
          />
        </td>
      </tr>
      <tr
        class="membership-request-row"
        id="table-row-1"
      >
        <td
          class=""
          colspan="1"
        >
          Jane Smith
        </td>
        <td
          class=""
          colspan="1"
        >
          11/27/2022
        </td>
        <td
          class=""
          colspan="1"
        >
          <div
            class="cf-dropdown"
            data-css-fp8ian=""
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              aria-label="Select action"
              aria-pressed="false"
              class="cf-dropdown-trigger usa-button usa-button-secondary"
              data-css-x1q03x=""
              id="dropdown-button"
              role="dropdown-button"
            >
              Select action
            </button>
          </div>
        </td>
        <td
          class=""
          colspan="1"
        >
          <button
            aria-expanded="false"
            aria-label="Show note"
            class="usa-accordion-button"
            style="background-color: inherit;"
          />
        </td>
      </tr>
      <tr
        class="membership-request-row"
        id="table-row-2"
      >
        <td
          class=""
          colspan="1"
        >
          William Brown
        </td>
        <td
          class=""
          colspan="1"
        >
          12/01/2022
        </td>
        <td
          class=""
          colspan="1"
        >
          <div
            class="cf-dropdown"
            data-css-fp8ian=""
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              aria-label="Select action"
              aria-pressed="false"
              class="cf-dropdown-trigger usa-button usa-button-secondary"
              data-css-x1q03x=""
              id="dropdown-button"
              role="dropdown-button"
            >
              Select action
            </button>
          </div>
        </td>
        <td
          class=""
          colspan="1"
        />
      </tr>
      <tr
        class="membership-request-row"
        id="table-row-3"
      >
        <td
          class=""
          colspan="1"
        >
          Emma Wilson
        </td>
        <td
          class=""
          colspan="1"
        >
          11/20/2022
        </td>
        <td
          class=""
          colspan="1"
        >
          <div
            class="cf-dropdown"
            data-css-fp8ian=""
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              aria-label="Select action"
              aria-pressed="false"
              class="cf-dropdown-trigger usa-button usa-button-secondary"
              data-css-x1q03x=""
              id="dropdown-button"
              role="dropdown-button"
            >
              Select action
            </button>
          </div>
        </td>
        <td
          class=""
          colspan="1"
        />
      </tr>
      <tr
        class="membership-request-row"
        id="table-row-4"
      >
        <td
          class=""
          colspan="1"
        >
          Micheal Johnson
        </td>
        <td
          class=""
          colspan="1"
        >
          11/30/2022
        </td>
        <td
          class=""
          colspan="1"
        >
          <div
            class="cf-dropdown"
            data-css-fp8ian=""
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              aria-label="Select action"
              aria-pressed="false"
              class="cf-dropdown-trigger usa-button usa-button-secondary"
              data-css-x1q03x=""
              id="dropdown-button"
              role="dropdown-button"
            >
              Select action
            </button>
          </div>
        </td>
        <td
          class=""
          colspan="1"
        />
      </tr>
    </tbody>
    <tfoot />
  </table>
</div>
`;

exports[`MembershipRequestTable renders the default state with no requests correctly 1`] = `
<div>
  <h2>
    View 0 pending requests
  </h2>
  <table
    class="usa-table-borderless membership-request-table"
  >
    <thead
      class=""
    >
      <tr>
        <th
          class=""
          scope="col"
        >
          <span>
            User name
          </span>
        </th>
        <th
          class=""
          scope="col"
        >
          <span>
            Date requested
          </span>
        </th>
        <th
          class=""
          scope="col"
        >
          <span>
            Actions
          </span>
        </th>
        <th
          class=""
          scope="col"
        >
          <span />
        </th>
      </tr>
    </thead>
    <tbody
      class=""
    />
    <tfoot />
  </table>
</div>
`;
